llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -9962,6 +9962,18 @@ SDValue DAGCombiner::visitSHL(SDNode *N) {
     if (SDValue NewSHL = visitShiftByConstant(N))
       return NewSHL;
 
+  // fold (shl X, cttz(Y)) -> (mul (Y & -Y), X) if cttz is unsupported on the
+  // target.
+  if ((N1.getOpcode() == ISD::CTTZ || N1.getOpcode() == ISD::CTTZ_ZERO_UNDEF) &&
+      N1.hasOneUse() && !TLI.isOperationLegalOrCustom(ISD::CTTZ, VT) &&
```

Comment:
You check cttz||cttz_zero_undef but hardcode the opcode in the legality check. Should you check for getOpcode's legality instead?

Reply:
I hardcode the opcode to avoid introducing regressions on rv64+zbb :(
Do you have better solution?

Reply:
Right condition might be isLegalOrCustom(CTTZ||CTTZ_ZERO_UNDEF)

Reply:
> Right condition might be isLegalOrCustom(CTTZ||CTTZ_ZERO_UNDEF)

Unfortunately it doesn't work :(

Reply:
Is it suitable to add a TLI hook?